### PR TITLE
proxy: switch to leaky bucket

### PR DESCRIPTION
**Pull request by conradludgate**
## Problem

The current bucket based rate limiter is not very intuitive and has some bad failure cases.

## Summary of changes

Switches from fixed interval buckets to leaky bucket impl. A single bucket per endpoint,
drains over time. Drains by checking the time since the last check, and draining tokens en-masse. Garbage collection works similar to before, it drains a shard (1/64th of the set) every 2048 checks, and it only removes buckets that are empty.

To be compatible with the existing config, I've faffed to make it take the min and the max rps of each as the sustained rps and the max bucket size which should be roughly equivalent.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: proxy/src/rate_limiter/limiter.rs [hunk 1]
```diff
@@ -293,7 +297,7 @@ mod tests {
             .map(|s| s.parse().unwrap())
             .collect();
         RateBucketInfo::validate(&mut rates).unwrap();
-        let limiter = EndpointRateLimiter::new(rates);
+        let limiter = BucketRateLimiter::new(rates);
```
Comment (stradig) [inline proxy/src/rate_limiter/limiter.rs:300]: Any reason not to use WakeupComputeRateLimiter here?

**Review by stradig — APPROVED** [review]
Very nice! Much easier to understand than the old limiter.

LGTM minus some nits.

File: proxy/src/rate_limiter/leaky_bucket.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        let mut entry = self.map.entry(key).or_insert_with(|| LeakyBucketState {
+            time: now,
+            filled: 0.0,
+        });
+
+        entry.check(&self.info, now, n as f64)
+    }
+
+    fn do_gc(&self, now: Instant) {
+        info!(
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/rate_limiter/leaky_bucket.rs:54]: It's one line for each 2048 requests. Seems fine

File: proxy/src/rate_limiter/limiter.rs (hunk 1 above)
Comment (conradludgate) [reply proxy/src/rate_limiter/limiter.rs:300]: Nope. Just missed it during the refactor

File: proxy/src/rate_limiter/leaky_bucket.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+use tracing::info;
+
+use crate::intern::EndpointIdInt;
+
+// Simple per-endpoint rate limiter.
+pub type EndpointRateLimiter = LeakyBucketRateLimiter<EndpointIdInt>;
+
+pub struct LeakyBucketRateLimiter<Key> {
+    map: DashMap<Key, LeakyBucketState, RandomState>,
+    info: LeakyBucketConfig,
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/rate_limiter/leaky_bucket.rs:19]: It's the same name the old impl used. I could rename it

[the remark this answers is not in the document]
Comment (stradig) [reply proxy/src/rate_limiter/leaky_bucket.rs:19]: Not too important. Just seemed like an odd name. Your call.

File: proxy/src/rate_limiter/leaky_bucket.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stradig) [reply proxy/src/rate_limiter/leaky_bucket.rs:54]: Fine with me, then.

**Pull request merged**